### PR TITLE
Followup to 2949 moved and edited the nuage stuff

### PR DESCRIPTION
**Pull request by bfallonf**
#2949

I started to edit the content from the above PR, but soon realised that the location and context might not be perfect.

@aheslin Do you have any suggestions?

Aside from the sentence edits...

I've moved the stuff from the nuage.adoc topic into the additional_concepts/networking.adoc topic. I think it's a better house instead of a brand new. I've moved the stuff from the configuring_nuagesdn.adoc file into the configuring_sdn.adoc file. I think it's more efficient to have one file on the two different methods.

I've not deleted the new files added in from #2949, but just copied the content. I did this in case anyone has a better suggestion. @aheslin I hope this is not confusing...

Additionally, @xltian would it be possible to QA the one procedure in the massive content here? I'll mark where it starts below.

Thanks!

cc: @vikram-redhat

### CODE REVIEW

File: install_config/configuring_sdn.adoc [hunk 1]
```diff
@@ -164,6 +164,147 @@ Check VNIDs by running:
...
+
+==== Operations workflow
+
+This workflow is used by operations teams doing production roll outs of their applications.
+
+In this workflow, the network and security policies are first configured on the VSD in accordance with the rules set by the organization to deploy applications. Administrative users can potentially create multiple zones and subnets and map them to the same project using labels. While spinning up the pods, the user can use the Nuage Labels to specify what network a pod needs to attach to and what network policies need to be applied to it. This allows for sophisticated deployments where inter- and intra-project traffic can be controlled in a fine-grained manner. For example, inter-project communication is enabled on a project by project basis. This may be used to connect projects to common services that are deployed in a shared project.
+
+[[integration-components]]
+
+=== Installation
```
Comment (bfallonf) [inline install_config/configuring_sdn.adoc]: @xltian The procedure starts below. Thanks!

File: install_config/configuring_sdn.adoc [hunk 2]
```diff
@@ -164,6 +164,147 @@ Check VNIDs by running:
...
+ plugin_rpm=http://location_of_rpm_server/openshift/RPMS/x86_64/vsp-openshift-4.0.X1830.el7.centos.x86_64.rpm
+
+ # Optional parameters
+ nuage_interface_mtu=1460
+ nuage_master_adminusername=admin
+ nuage_master_adminuserpasswd=admin
+ nuage_master_cspadminpasswd=admin
+ nuage_openshift_monitor_log_dir=/var/log/nuage-openshift-monitor
+
+ # Required for brownfield install (where a {product-title} cluster exists without Nuage as the networking plugin)
```
Comment (abhat) [inline install_config/configuring_sdn.adoc]: the {product-title} macro is not expanding when in a code snippet tag. Can we just change this to Openshift?

File: install_config/configuring_sdn.adoc [hunk 3]
```diff
@@ -164,6 +164,147 @@ Check VNIDs by running:
...
+ nuage_dockker_bridge=lbr0
```
Comment (abhat) [inline install_config/configuring_sdn.adoc]: there is a typo - should be nuage_docker_bridge.

File: install_config/configuring_sdn.adoc [hunk 4]
```diff
@@ -164,6 +164,147 @@ Check VNIDs by running:
...
+ nuage_openshift_monitor_rest_server_port=9443
+ 
+ # Optional parameters
+ nuage_interface_mtu=1460
+ nuage_master_adminusername='admin's user-name'
+ nuage_master_adminuserpasswd='admin's password'
+ nuage_master_cspadminpasswd='csp admin password'
+ nuage_openshift_monitor_log_dir=/var/log/nuage-openshift-monitor
+
+ # Required for brownfield install (where a {product-title} cluster exists without Nuage as the networking plugin)
```
Comment (abhat) [inline install_config/configuring_sdn.adoc]: Same comment here as before. the {product-title} macro doesn't expand in a code snippet.

File: install_config/configuring_sdn.adoc [hunk 5]
```diff
@@ -164,6 +164,147 @@ Check VNIDs by running:
...
+ nuage_dockker_bridge=lbr0
```
Comment (abhat) [inline install_config/configuring_sdn.adoc]: typo `nuage_dockker_bridge` vs. `nuage_docker_bridge`

**Comment by abhat** [comment]
Linking #2949

**Comment by abhat** [comment]
@bfallonf I don't see the Nuage content in the latest origin documentation. Can you take a look?

**Comment by adellape** [comment]
@abhat I confirmed that the Nuage content wasn't there in the published Origin docs, even though it was supposed to be (https://github.com/openshift/openshift-docs/pull/2949 was merged Sept 29). I looked into it with @nhr and he found the bug in our build tool, which he fixed via https://github.com/redhataccess/ascii_binder/pull/68.

The Origin docs have since been rebuilt, and the Nuage content is now available as expected:

https://docs.openshift.org/latest/architecture/additional_concepts/nuagesdn.html
https://docs.openshift.org/latest/install_config/configuring_nuagesdn.html

Apologies for the issue!

FYI @vikram-redhat

**Comment by bfallonf** [comment]
OK. Sounds like this could be merged now that the original PR has merged.

@adellape @ahardin-rh @aheslin @vikram-redhat I fixed the conflicts on this PR. Maybe a peer review is in order? Unless there's comments on the arranging I did here (See comment 1)?

**Comment by bfallonf** [comment]
The Nuage stuff found itself in the docs. This hasn't merged yet, so there's not been a follow-up.

Vikram brought this up with me today and I went through it again and made some changes. 

@ahardin-rh @bmcelvee @adellape Any chance for a peer review please?

File: architecture/additional_concepts/networking.adoc [hunk 6]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+
+.Nuage VSP Building Blocks
+image::nuage_terminology.png[Nuage VSP Building Blocks]
+
+ . Domains: An organization contains one or more domains. A domain is a single "Layer 3" space. In standard networking terminology, a domain maps to a VRF instance.
+ . Zones: Zones are defined under a domain. A zone does not map to anything on the network directly, but instead acts as an object with which policies are associated such that all endpoints in the zone adhere to the same set of policies.
+ . Subnets: Subnets are defined under a zone. A subnet is a specific Layer 2 subnet within the domain instance. A subnet is unique and distinct within a domain, that is, subnets within a Domain are not allowed to overlap or to contain other subnets in accordance with the standard IP subnet definitions.
+ . VPorts: A VPort is a new level in the domain hierarchy, intended to provide more granular configuration. In addition to containers and VMs, VPorts are also used to attach Host and Bridge Interfaces, which provide connectivity to Bare Metal servers, Appliances, and Legacy VLANs.
+ . Policy Group: Policy Groups are collections of VPorts. 
+
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: Missing section ID/anchor here

File: architecture/additional_concepts/networking.adoc [hunk 7]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+A pod spawning in {product-title} translates to a virtual port being created in
+VSP. The *vsp-openshift* plug-in interacts with the VRS and gets a policy for
+that virtual port from the VSD via the VSC. Policy Groups are supported to group
+multiple pods together that must have the same set of policies applied to them.
+Currently, pods can only be assigned to policy groups using the
+xref:nuagesdn.adoc#operations-workflow[operations workflow] where a policy group
+is created by the administrative user in VSD. The pod being a part of the policy
+group is specified by means of nuage.io/policy-group label in the specification
+of the pod.
+
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: Missing section ID/anchor

File: architecture/additional_concepts/networking.adoc [hunk 8]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+{product-title} deploys a software-defined networking (SDN) approach for
+connecting pods in an {product-title} cluster. The {product-title} SDN connects
+all pods across all node hosts, providing a unified cluster network.
 
 {product-title} SDN is automatically installed and configured as part of the
-Ansible-based installation procedure.  Further administration should not
-be required; however, xref:sdn.adoc#architecture-additional-concepts-sdn[further details on the design and
-operation of {product-title} SDN] are provided for those who are curious or
-need to troubleshoot problems.
+Ansible-based installation procedure. See the
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: Maybe remove "See the SDN section for more information" and make "Ansible-based installation procedure" a link instead (if that is what is discussed there).

File: architecture/additional_concepts/networking.adoc [hunk 9]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
 ifdef::openshift-enterprise,openshift-origin[]
 [[nuage-sdn]]
 === Nuage SDN for {product-title}
 
-Nuage Networks' SDN solution based on it's link:http://www.nuagenetworks.net/products/virtualized-services-platform/[Virtualized Services Platform] delivers
-highly scalable policy based overlay networking for pods in an {product-title} cluster. Nuage SDN can be installed and configured as a part of the
-Ansible-based installation procedure. More information about the xref:nuagesdn.adoc#architecture-additional-concepts-nuagesdn[design and integration] is provided for a deep-dive. The xref:../../install_config/configuring_nuagesdn.adoc#install-config-configuring-nuage-sdn[installation section] will have detailed information about how to install and deploy {product-title} with Nuage SDN.
+link:http://www.nuagenetworks.net/products/virtualized-services-platform[Nuage
+Networks'] SDN solution delivers highly scalable, policy-based overlay
+networking for pods in an {product-title} cluster. Nuage SDN can be installed
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: Maybe make Nuage SDN a link to xref:../../architecture/additional_concepts/nuagesdn.adoc#architecture-additional-concepts-nuagesdn here, then remove the "See the Nuage SDN section for more information" blurb. Then, you can continue with "See the Advance Installation section...." blurb.

File: architecture/additional_concepts/networking.adoc [hunk 10]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+plane.
+
+Nuage uses overlays to provide policy-based networking between {product-title}
+and other environments consisting of VMs and bare metal servers. The platform’s
+real-time analytics engine enables visibility and security monitoring for
+{product-title} applications.
+
+Nuage VSP integrates with {product-title} to allows business applications to be
+quickly turned up and updated by removing the network lag faced by DevOps teams. 
+.Nuage VSP Integration with {product-title}
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: Missing space between .Nuage VSP... and the previous line

File: architecture/additional_concepts/networking.adoc [hunk 11]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+particular project can span multiple nodes in {product-title}.
+
+A pod spawning in {product-title} translates to a virtual port being created in
+VSP. The *vsp-openshift* plug-in interacts with the VRS and gets a policy for
+that virtual port from the VSD via the VSC. Policy Groups are supported to group
+multiple pods together that must have the same set of policies applied to them.
+Currently, pods can only be assigned to policy groups using the
+xref:nuagesdn.adoc#operations-workflow[operations workflow] where a policy group
+is created by the administrative user in VSD. The pod being a part of the policy
+group is specified by means of nuage.io/policy-group label in the specification
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: `nuage.io/policy-group` label

File: architecture/additional_concepts/networking.adoc [hunk 12]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+
+. nuage-openshift-monitor
+. vsp-openshift plugin
+
+[[nuage-openshift-monitor]]
+
+=== nuage-openshift-monitor
+
+The nuage-openshift-monitor is a service that monitors the {product-title} API
+server for creation of projects, services, users, user-groups etc.
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: comma before etc. (or just spell our "and so on")

File: architecture/additional_concepts/networking.adoc [hunk 13]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+Nuage VSP integrates with {product-title} using two main components:
+
+. nuage-openshift-monitor
+. vsp-openshift plugin
+
+[[nuage-openshift-monitor]]
+
+=== nuage-openshift-monitor
+
+The nuage-openshift-monitor is a service that monitors the {product-title} API
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: *nuage-openshift-monitor*

File: architecture/additional_concepts/networking.adoc [hunk 14]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+server for creation of projects, services, users, user-groups etc.
+
+[NOTE]
+=====
+In case of a Highly Available (HA) {product-title} cluster with multiple
+masters, nuage-openshift-monitor process runs on all the masters independently
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: *nuage-openshift-monitor*

File: architecture/additional_concepts/networking.adoc [hunk 15]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+objects by exercising the VSD REST API to map {product-title} constructs to VSP
+constructs. Each cluster instance maps to a single domain in Nuage VSP. This
+allows a given enterprise to potentially have multiple cluster installations -
+one per domain instance for that Enterprise in Nuage. Each {product-title}
+project is mapped to a zone in the domain of the cluster on the Nuage VSP.
+Whenever the nuage-openshift-monitor sees an addition or deletion of the
+project, it instantiates a zone using the VSDK APIs corresponding to that
+project and allocates a block of subnet for that zone. Additionally, the
+nuage-openshift-monitor also creates a network macro group for this project.
+Likewise, whenever the nuage-openshift-monitor sees an addition/deletion of a
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: addition or deletion

File: architecture/additional_concepts/networking.adoc [hunk 16]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+service, it creates a network macro corresponding to the service IP and assigns
+that network macro to the network macro group for that project (user provided
+network macro group using labels is also supported) to enable communication to
+that service.
+
+For the developer workflow, all pods that are created within the zone get IPs
+from that subnet pool. The subnet pool allocation and management is done by the
+nuage-openshift-monitor based on a couple of plug-in specific parameters in the
+master-config file. However the actual IP address resolution and vport policy
+resolution is still done by VSD based on the domain/zone that gets instantiated
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: domain or zone

File: architecture/additional_concepts/networking.adoc [hunk 17]
```diff
@@ -94,32 +97,166 @@ round-robin between each pod.
...
+the VSD constructs to map the pods into a specific zone/subnet as well as allow
+communication between OpenShift entities (ACL rules, policy groups, network
+macros, and network macro groups). Detailed description of how to use Nuage
+labels is provided in the link:http://support.alcatel-lucent.com[Nuage VSP
+Openshift Integration Guide].
+=====
+
+[[vsp-openshift-plugin]]
+
+=== vsp-openshift plug-in
```
Comment (ahardin-rh) [inline architecture/additional_concepts/networking.adoc]: Plug-in (in heading title)

File: install_config/configuring_nuagesdn.adoc [hunk 18]
```diff
@@ -10,16 +10,64 @@
...
+applications to be quickly turned up and updated by removing the network lag
+faced by DevOps teams. VSP supports different workflows with {product-title} in
+order to accommodate scenarios where users can choose ease-of-use or complete
+control using policy-based automation.
+
+*Developer Workflow*
+
+This workflow is used in developer environments and requires little input from
+the developer in setting up the networking. In this workflow, the
+nuage-openshift-monitor is responsible for creating the VSP constructs (Zone,
```
Comment (ahardin-rh) [inline install_config/configuring_nuagesdn.adoc]: unnecessary "the" before the service name

*nuage-openshift-monitor*

File: install_config/configuring_nuagesdn.adoc [hunk 19]
```diff
@@ -10,16 +10,64 @@
...
+example, inter-project communication is enabled on a project by project basis.
+This may be used to connect projects to common services that are deployed in a
+shared project.
+
+[[nuage-sdn-installation]]
+
+== Installation
+
+The VSP integration with {product-title} works for both virtual machines (VMs)
+and baremetal {product-title} installations.
```
Comment (ahardin-rh) [inline install_config/configuring_nuagesdn.adoc]: bare metal

File: install_config/configuring_nuagesdn.adoc [hunk 20]
```diff
@@ -10,16 +10,64 @@
...
+
+[[nuage-sdn-installation-single-master]]
 === Installation for single master
 
-In the Ansible nodes file, the following parameters need to be specified in order to setup Nuage VSP as the network plugin.
+In the Ansible nodes file, specify the following parameters in order to set up
+Nuage VSP as the network plugin:
```
Comment (ahardin-rh) [inline install_config/configuring_nuagesdn.adoc]: plug-in

File: install_config/configuring_nuagesdn.adoc [hunk 21]
```diff
@@ -50,13 +98,25 @@ In the Ansible nodes file, the following parameters need to be specified in orde
...
+load balancing solution, the default being HAProxy. The inventory file contains
+three master hosts, the nodes, an etcd server, and a host that functions as the
+HAProxy to balance the master API on all master hosts. The HAProxy host is
+defined in the [lb] section of the inventory file enabling Ansible to
+automatically install and configure HAProxy as the load balancing solution. 
 
-Nuage VSP integration in multi-master mode only supports the native HA configuration method described in this section. This can be combined with any load balancing solution, the default being HAProxy. In the inventory file, there are three master hosts, the nodes, an etcd server and a host that functions as the HAProxy to balance the master API on all master hosts. The HAProxy host is defined in the [lb] section of the inventory file enabling Ansible to automatically install and configure HAProxy as the load balancing solution. In the Ansible nodes file, the following parameters need to be specified in order to setup Nuage VSP as the network plugin:
+In the Ansible nodes file, the following parameters need to be specified in order to setup Nuage VSP as the network plugin:
```
Comment (ahardin-rh) [inline install_config/configuring_nuagesdn.adoc]: plug-in

File: architecture/additional_concepts/networking.adoc (hunk 8 above)
Comment (bfallonf) [inline architecture/additional_concepts/networking.adoc]: I thin this one is OK as is. It's linking to the more contextual sdn stuff. Linking to the install stuff would change the meaning.

**Comment by bfallonf** [comment]
Thanks @ahardin-rh I'll merge.

**Comment by bfallonf** [comment]
[rev_history]
|xref:../architecture/additional_concepts/networking.adoc#architecture-additional-concepts-networking[Additional Concepts -> Networking]
|Added information about Nuage SDN
%
[rev_history]
|xref:../install_config/configuring_nuagesdn.adoc#install-config-configuring-nuage-sdn[Configuring Nuage SDN]
|Added the Configuring Nuage SDN file.
%

**Pull request merged**